### PR TITLE
add StreamingIteratorMut::flatten

### PR DESCRIPTION
**Pull request by LorenzSchueler**
* add `StreamingIteratorMut::flatten`
* add `Flatten`
* add `impl StreamingIterator for Flatten`
* add `impl StreamingIteratorMut for Flatten`
* add test
* fix clippy warnings

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1264,6 +1277,61 @@ where
     }
 }
 
+/// A streaming iterator that flattens nested streaming iterators.
+#[derive(Debug)]
+pub struct Flatten<I, J> {
+    iter: I,
+    _inner: PhantomData<J>,
+}
```
Comment (cuviper) [inline src/lib.rs:1280]: I don't think we need the `J` parameter at all, and then you don't need to phantom anything.

(The standard library also only has `Flatten<I>`.)

File: src/lib.rs [hunk 2]
```diff
@@ -1264,6 +1277,61 @@ where
...
+where
+    I: StreamingIteratorMut<Item = J>,
+    J: StreamingIterator,
+{
+    type Item = J::Item;
+
+    #[inline]
+    fn advance(&mut self) {
+        loop {
+            if let Some(ref mut iter) = self.iter.get_mut() {
```
Comment (cuviper) [inline src/lib.rs]: `get_mut` documents:
> The behavior of calling this method before `advance` has been called is unspecified.

I think we'll need a `first: bool` flag for the first round, and then we should also not call `advance` any more after seeing `get_mut() -> None` or `is_done()`.

File: src/lib.rs [hunk 3]
```diff
@@ -1264,6 +1277,61 @@ where
...
+            Some(iter) => iter.is_done(),
+            None => true,
+        }
+    }
+
+    #[inline]
+    fn get(&self) -> Option<&Self::Item> {
+        self.iter.get().and_then(|iter| iter.get())
+    }
+}
```
Comment (cuviper) [inline src/lib.rs:1326]: Please also add `fold` for internal iteration -- probably in terms of `self.iter.fold_mut`.

File: src/lib.rs [hunk 4]
```diff
@@ -1264,6 +1277,61 @@ where
...
+impl<I, J> StreamingIteratorMut for Flatten<I, J>
+where
+    I: StreamingIteratorMut<Item = J>,
+    J: StreamingIteratorMut,
+{
+    #[inline]
+    fn get_mut(&mut self) -> Option<&mut Self::Item> {
+        self.iter.get_mut().and_then(J::get_mut)
+    }
+}
```
Comment (cuviper) [inline src/lib.rs:1347]: Please also add `fold_mut` for internal iteration.

File: src/lib.rs [hunk 5]
```diff
@@ -178,7 +178,7 @@ pub trait StreamingIterator {
         Self: Sized,
         F: FnMut(&Self::Item) -> bool,
     {
-        Filter { it: self, f: f }
+        Filter { it: self, f }
```
Comment (cuviper) [inline src/lib.rs]: I would rather the clippy changes be considered in a separate PR -- can you rebase without that commit?

**Review by cuviper — CHANGES_REQUESTED** [review]
This is clever to build `Flatten` on `StreamingIteratorMut`! That gives it a kind of superpower that the regular `FlatMap` can't match with just `identity`, since this can iterate the borrowed inner iterator in-place.

**Comment by LorenzSchueler** [comment]
Thanks for the review. I agree with your suggestions. Currently I am on vacation but I will follow up on this once I get back.
(In case you want to merge this earlier feel free to apply the changes yourself.)

**Comment by LorenzSchueler** [comment]
I opened #28 for the clippy warnings

File: src/lib.rs [hunk 6]
```diff
@@ -1264,6 +1278,85 @@ where
...
+
+    #[inline]
+    fn advance(&mut self) {
+        if self.first {
+            self.first = false;
+            self.iter.advance();
+        }
+        while let Some(iter) = self.iter.get_mut() {
+            if self.inner_first || !iter.is_done() {
+                self.inner_first = false;
```
Comment (cuviper) [inline src/lib.rs]: Why do we need `inner_first`? I think we can unconditionally `advance` the inner iterator here, just like `FlatMap` does. That can always be called at least once, and if that sees `is_done()`, then we're moving on to the next `self.iter.advance()` anyway.

Comment (LorenzSchueler) [reply src/lib.rs]: You are right. I somehow missed the fact that after we advance the outer iterator (`self.iter`) we have not yet called advance on the new inner one and are therefor able to call `advance` on it unconditionally.

File: src/lib.rs [hunk 7]
```diff
@@ -1264,6 +1277,80 @@ where
...
+#[derive(Debug)]
+pub struct Flatten<I> {
+    iter: I,
+    first: bool,
+}
+
+impl<I, J> StreamingIterator for Flatten<I>
+where
+    I: StreamingIteratorMut<Item = J>,
+    for<'a> J: StreamingIterator + 'a,
```
Comment (cuviper) [inline src/lib.rs]: This creates an awkward lifetime error in `fn get` on anything before Rust 1.63, which I believe was due to the migrate borrowck mode, removed in rust-lang/rust#95565. However, it works on older versions if we use `I::Item` without any lifetime constraints, rather than a separate `J` parameter.

File: src/lib.rs [hunk 8]
```diff
@@ -597,6 +597,19 @@ pub trait StreamingIteratorMut: StreamingIterator {
     {
         MapDerefMut { it: self, f }
     }
+
+    /// Creates an iterator which flattens nested streaming iterators.
+    #[inline]
+    fn flatten(self) -> Flatten<Self>
+    where
+        Self: Sized,
+        Self::Item: StreamingIterator + Sized,
```
Comment (cuviper) [inline src/lib.rs]: This doesn't need to be `Sized` anymore -- I think that was only because you didn't have `J: ?Sized` before, but that's gone now anyway.

File: src/lib.rs [hunk 9]
```diff
@@ -1264,6 +1277,80 @@ where
...
+        if self.first {
+            self.first = false;
+            self.iter.advance();
+        }
+        while let Some(iter) = self.iter.get_mut() {
+            iter.advance();
+            if !iter.is_done() {
+                break;
+            }
+            self.iter.advance(); // since we got Some, self.iter is not done and can be advanced
```
Comment (cuviper) [inline src/lib.rs:1300]: The pre-NLL borrowck didn't like this, mutating `self.iter` while `get_mut()` was lexically alive. I moved the crate to 2018 edition in #29, so now we do get NLL here and it's happy.

**Review by cuviper — APPROVED** [review]
I rebased and added a couple more fixes for the things below. If that looks good to you, then I think we're all set!

**Comment by LorenzSchueler** [comment]
looks good to me :)
